File: PeakFinding/PeakFinder.py
import copy

def findPeak(arr):
    n = int(len(arr) / 2);

    if n > 0 and arr[n] < arr[n - 1]:
        return findPeak(arr[0:n])
    elif n + 1 < len(arr) and arr[n] < arr[n + 1]:
        return findPeak(arr[n:len(arr)])
    else:
        return True, arr[n]

def findPeak_2D(arr2D, row_start, row_end, col_start, col_end):
    col = int(col_start + (col_end - col_start) / 2)
    row = findGlobalMax(arr2D, col)

    if col > 0 and arr2D[row][col] < arr2D[row][col - 1]:
        return findPeak_2D(arr2D, row_start, row_end, col_start, col)
    elif col + 1 < len(arr2D) and arr2D[row][col] < arr2D[row][col + 1]:
        return findPeak_2D(arr2D, row_start, row_end, col, col_end)
    else:
        return True, arr2D[row][col]

def findGlobalMax(arr2D, col):
    ret = None
    
    for index in range(len(arr2D)):
        if ret is None or arr2D[index][col] > arr2D[ret][col]:
            ret = index

    return ret


nums = [
		[10, 10, 8, 10],
		[11, 12, 13, 14],
		[21, 11, 9, 15],
		[20, 19, 17, 16]
	]

found, val = findPeak_2D(nums, 0, len(nums), 0, len(nums[0]))

print(found, val)

File: PeakFinding/PeakFinder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace PeakFinder
{ 
    class Program
    {
        private static Tuple<bool, int> findPeak(int[] arr, int start, int end)
        {
            int n = start + (end - start) / 2;

            if (n - 1 > 0 && arr[n] < arr[n - 1])
            {
                return findPeak(arr, start, n);
            }
            else if (n + 1 < end && arr[n] < arr[n + 1])
            {
                return findPeak(arr, n, end);
            }
            else
                return new Tuple<bool, int>(true, arr[n]);
        }

        private static int findGlobalMax(int[,] arr2D, int col)
        {
            int ret = 0;

            for(int row = 0; row < arr2D.GetLength(0); ++row)
            {
                if(arr2D[row, col] > arr2D[ret, col])
                {
                    ret = row;
                }
            }

            return ret;
        }

        private static Tuple<bool, int> findPeak_2D(int[,] arr2D, int start_row, int end_row, int start_col, int end_col)
        {
            int col = start_col + (end_col - start_col) / 2;
            int row = findGlobalMax(arr2D, col);

            if(start_col <= col - 1 && arr2D[row, col] < arr2D[row, col - 1])
            {
                return findPeak_2D(arr2D, start_row, end_row, start_col, col);
            }
            else if(end_col > col + 1 && arr2D[row, col] < arr2D[row, col + 1])
            {
                return findPeak_2D(arr2D, start_row, end_row, col, end_col);
            }
            else
            {
                return new Tuple<bool, int>(true, arr2D[row, col]);
            }
        }

        static void Main(string[] args)
        {
            //int[] arr = { 1, 9, 5, 4, 3, 6, 7 };
            //Tuple<bool, int> result = findPeak(arr, 0, arr.Length);

            int[,] arr2D =
            {
                { 10, 10, 8, 10 },
                { 11, 12, 13, 14 },
                { 21, 11, 9, 15 },
                { 20, 19, 17, 16 }
            };

            var result = findPeak_2D(arr2D, 0, arr2D.GetLength(0), 0, arr2D.GetLength(1));

            System.Diagnostics.Debug.WriteLine("Found peak: {0}, and the peak value is: {1}", result.Item1, result.Item2);
        }
    }
}
